NOTAM Integration

yorickreum opened the pull request:
This PR is a new version of https://github.com/XCSoar/XCSoar/pull/1921, which I closed as I rebased several times and it is now overspammed with CodeRabbits comments.

The PR adds NOTAM support to XCSoar, displaying NOTAMs as airspaces on the moving map and as a reference list in the settings. The goal is to improve situational awareness and flight safety.

Enable via: `Settings → Map Display → Airspace → NOTAM`

Features:

- Displays NOTAMs as map airspaces
- Settings list for review/verification
- Integrated into existing airspace pipeline
- Filtering with sensible defaults (reduce alert fatigue)
- NOTAMs rendered blue/grayish, not warning red
- Automatic updates (after change of location or a certain time)
- and incremental (delta) updating supported

As data source it's now using an FAA NOTAM proxy based on the new OAuth2 FAA API, the code for that is in https://github.com/yorickreum/xcsoar-notam-proxy. I updated the wrapper from @Akaflieg-Freiburg to the new FAA API and added the delta/incremental update functionality.

I'll leave out the .po files for translation in this PR to keep the CodeRabbit focused on the relevant code changes, but I have them ready for a separate PR in https://github.com/yorickreum/XCSoar/tree/notam-delta-po.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full NOTAM integration: download/cache NOTAMs, convert to airspaces, NOTAM list & details dialogs, NOTAM settings panel, map NOTAM labels with clustering, manual refresh and auto-refresh.

* **Bug Fixes / Reliability**
  * Safer UTF‑8 handling, stricter input validation, resilient caching/merge/retry, expanded error handling, improved ACK / Ack‑Day behavior, and rendering/label overlap fixes.

* **Tests**
  * New tests for NOTAM parsing, caching/metadata, filtering, and warning/ack logic.

* **Documentation**
  * NEWS entry describing NOTAM integration.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/XCSoar/XCSoar/pull/2018?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
src/NOTAM,build/libnotam.mk: Implement fetching, filtering and displaying of NOTAMs

Adds NOTAM (Notice to Air Missions) support with automatic fetching from APIs
based on current location with a configurable radius and interval refresh cycle.

Features:
- Automatic background fetching with configurable refresh intervals (0-240 min)
- Real-time location-based queries with distance threshold checks
- Multi-criteria filtering:
  * IFR-only traffic filtering
  * Time-based (show only currently effective NOTAMs)
  * Radius-based (filter by coverage area in user units)
  * Q-code prefix exclusion (customizable hide list)
- Settings panel with live filter statistics showing filtered counts per criterion
- Manual "Update Now" that applies settings immediately
- NOTAM list dialog with:
  * Status badges (Active/Expired/Future with countdown)
  * Last update timestamp and distance from fetch location
  * Visual filtering indicators on each NOTAM
  * Total and visible counts after filtering

Technical:
- Async HTTP fetching with Co::InvokeTask and automatic retry
- Cache invalidation based on location movement and time
- Integration with existing airspace settings infrastructure

Commit:
src/Dialogs/dlgAirspaceWarnings: Add Details button

yorickreum commented on src/NetComponents.hpp:
```diff
@@ -24,10 +26,12 @@ struct NetComponents {
   NetComponents(EventLoop &event_loop, CurlGlobal &curl,
-                const TrackingSettings &tracking_settings) noexcept;
+                const TrackingSettings &tracking_settings,
+                const NOTAMSettings &notam_settings) noexcept;
```
`HAVE_HTTP=n` currently is not supported by the codebase, properly addressed in https://github.com/XCSoar/XCSoar/pull/2004.

yorickreum commented on darwin/XCSoar.xcodeproj/project.pbxproj:
```diff
@@ -19730,6 +19761,7 @@
 		5CBFEDA72D742FA600F05795 /* src */ = {
 			isa = PBXGroup;
 			children = (
+				5C1532E82E92F15E00168360 /* NOTAM */,
```
Not a problem, as we don't use the XCode build system. (We use XCode only with make as an external build system.)

lordfolken commented:
> I'd like #1979 to be implemented before we merge this in.

I see your point. I agree. With NOTAMs it gets even more tricky.

lordfolken committed:
Renderer/AirspacePreviewRenderer: Restore list text color after GDI preview fill

UnprepareFill() reset text color to black, which broke dark-mode list rows after drawing the airspace preview swatch. Save GetTextColor() before PrepareFill and pass it to UnprepareFill on GDI builds. Update airspace settings list and cross-section call sites.

lordfolken committed:
Engine/Airspace/AirspaceWarningManager: Preserve NOTAM day-ack across refresh

NOTAM airspace objects are recreated when the NOTAM list updates, so warnings no longer matched the old AbstractAirspace pointer and Ack day appeared lost. Track day acknowledgement by NOTAM number (GetStationName()) in addition to the per-warning flag.

Commit:
NetComponentsStub: match NetComponents constructor signature

Commit:
MapWindow,Profile,Renderer: Add gated NOTAM map labels

Commit:
Dialogs: Add NOTAM briefing disclaimer

Commit:
Airspace/ProtectedAirspaceWarningManager: drop noexcept from acknowledge wrappers

yorickreum commented:
The CodeRabbit comments don't end...

yorickreum commented on src/NOTAM/NOTAM.hpp:
```diff
@@ -0,0 +1,121 @@
+   * Check if this NOTAM affects the given location and altitude
+   */
+  bool AffectsLocation(double longitude, double latitude,
+                       double altitude_meters = -1.0) const noexcept;
```
Not declared anymore.

yorickreum commented on src/Dialogs/Settings/Panels/NOTAMConfigPanel.cpp:
```diff
@@ -0,0 +1,392 @@
+    } else {
+      UpdateFilterCounts();
+      ShowMessageBox(_("No valid location."), _("NOTAM"),
+                     MB_OK | MB_ICONEXCLAMATION);
```
Fixed?

yorickreum commented on src/NOTAM/NOTAMGlue.cpp:
```diff
@@ -0,0 +1,1665 @@
+NOTAMGlue::LoadNOTAMsInternal(GeoPoint location)
+{
+  LogFormat("NOTAM: Starting LoadNOTAMsInternal for location %.6f,%.6f", 
+            location.latitude.Degrees(), location.longitude.Degrees());
```
Fixed?

yorickreum commented on src/NOTAM/NOTAMGlue.hpp:
```diff
@@ -0,0 +1,241 @@
+#include <ctime>
+#include <optional>
+
+class CurlGlobal;
+class Airspaces;
+class OperationEnvironment;
+struct NOTAMImpl; // Forward declaration for pimpl
+
+/**
+ * Listener interface for NOTAM updates
+ */
+class NOTAMListener {
+public:
+  virtual ~NOTAMListener() = default;
+  
+  /**
+   * Called when NOTAMs have been successfully loaded and filtered
+   */
+  virtual void OnNOTAMsUpdated() noexcept = 0;
+};
```
Should be okay now?

yorickreum commented on src/Renderer/AirspacePreviewRenderer.hpp:
```diff
@@ -17,7 +18,11 @@ namespace AirspacePreviewRenderer
+   * Restore GDI text drawing after #PrepareFill. Pass the text color that
+   * was active before #PrepareFill (e.g. from Canvas::GetTextColor()).
+   */
+  void UnprepareFill(Canvas &canvas, Color text_color) noexcept;
```
Fixed now?

yorickreum replied:
Now better?

yorickreum commented on src/Engine/Airspace/AirspaceWarningManager.hpp:
```diff
@@ -43,6 +45,14 @@ class AirspaceWarningManager {
+   * day" for NOTAM is also keyed by NOTAM number (#GetStationName()) so it
+   * survives updates.
+   */
+  std::unordered_set<std::string> notam_day_ack_by_station;
```
I think this is fixed?

yorickreum commented:
@coderabbitai resume

yorickreum commented on src/Engine/Airspace/AirspaceWarningManager.cpp:
```diff
@@ -419,28 +453,40 @@ AirspaceWarningManager::Acknowledge(ConstAirspacePtr airspace) noexcept
 
 void
 AirspaceWarningManager::AcknowledgeWarning(ConstAirspacePtr airspace,
-                                           const bool set) noexcept
+                                           const bool set)
 {
   GetWarning(std::move(airspace)).AcknowledgeWarning(set);
 }
 
 void
 AirspaceWarningManager::AcknowledgeInside(ConstAirspacePtr airspace,
-                                          const bool set) noexcept
+                                          const bool set)
 {
   GetWarning(std::move(airspace)).AcknowledgeInside(set);
 }
 
 void
 AirspaceWarningManager::AcknowledgeDay(ConstAirspacePtr airspace,
-                                       const bool set) noexcept
+                                       const bool set)
 {
+  if (const char *key = NotamDayAckKey(*airspace); key != nullptr) {
+    if (set)
+      notam_day_ack_by_station.emplace(key);
+    else
+      notam_day_ack_by_station.erase(key);
+  }
+
   GetWarning(std::move(airspace)).AcknowledgeDay(set);
 }
 
 bool
 AirspaceWarningManager::GetAckDay(const AbstractAirspace &airspace) const noexcept
 {
+  if (const char *key = NotamDayAckKey(airspace);
+      key != nullptr &&
+      notam_day_ack_by_station.find(key) != notam_day_ack_by_station.end())
+    return true;
```
The Ack Session behaviour fits existing airspace ack behaviour and is sufficient, if not prefered.

yorickreum commented on src/Dialogs/Settings/Panels/NOTAMConfigPanel.cpp:
```diff
@@ -0,0 +1,453 @@
+
+  RowFormWidget::Hide();
+}
+
+void
+NOTAMConfigPanel::OnUpdateButton() noexcept
+{
+  LogFormat("NOTAM: Manual update triggered from settings panel");
+#ifdef HAVE_HTTP
+  // Save current settings first so the update uses the new values
+  bool dummy_changed = false;
+  Save(dummy_changed);
+
+  const bool notam_enabled =
+    CommonInterface::GetComputerSettings().airspace.notam.enabled;
+
+  if (net_components && net_components->notam && notam_enabled) {
+    const auto &basic = CommonInterface::Basic();
+    if (!basic.location.IsValid()) {
+      UpdateFilterCounts();
+      ShowMessageBox(_("No valid location."), _("NOTAM"),
+                     MB_OK | MB_ICONEXCLAMATION);
+      return;
+    }
+
+    net_components->notam->ResetFetchFailureNotification();
+
+    if (net_components->notam->ForceUpdateLocation(basic.location)) {
+      ShowLoadingStatus();
+      net_components->notam->MarkManualRefreshRequested();
+    }
```
Is this resolved now?

yorickreum commented:
@coderabbitai resume

yorickreum commented on src/Dialogs/MapItemListDialog.cpp:
```diff
@@ -28,8 +28,11 @@
 #include "Terrain/RasterTerrain.hpp"
 #include "Protection.hpp"
+#include "LogFile.hpp"
+#include "../Message.hpp"
```
`#include "Dialogs/Message.hpp"` is the wrong file, now includuding `<Message.hpp>`

yorickreum commented:
@coderabbitai full review

yorickreum commented on src/NOTAM/Client.cpp:
```diff
@@ -0,0 +1,626 @@
+
+  LogFmt("NOTAM: Fetching NOTAMs within {:.1f} NM radius", radius_nm);
+  LogDebug("NOTAM: Requesting from {} with radius {:.1f} NM",
+           base_url, radius_nm);
```
The base URL is public in the source code anyways?

yorickreum replied:
Frankely I think this is more than okay at least for now.

yorickreum commented on src/NOTAM/Client.cpp:
```diff
@@ -0,0 +1,626 @@
+                    const GeoPoint &location, ProgressListener &progress,
+                    const KnownMap *known)
+{
+  const char *const base_url = settings.api_base_url.c_str();
+  if (base_url[0] == '\0') {
+    LogFmt("NOTAM: API base URL is empty");
+    co_return NOTAMResponse{};
+  }
```
base url is public in the source code anyways?

yorickreum commented on src/Dialogs/Settings/Panels/NOTAMConfigPanel.cpp:
```diff
@@ -0,0 +1,471 @@
+    if (net_components != nullptr && net_components->notam != nullptr &&
+        data_components != nullptr && data_components->airspaces != nullptr &&
+        computer.notam.enabled) {
+      if (radius_changed && !saving_for_manual_update) {
+        const auto &basic = CommonInterface::Basic();
+        if (basic.location.IsValid()) {
+          net_components->notam->ForceUpdateLocation(basic.location, true);
+        }
```
I think this is quite an edge case and would add quite some unnecessary complexity?

yorickreum commented on src/Dialogs/Airspace/dlgAirspaceDetails.cpp:
```diff
@@ -110,21 +162,56 @@ AirspaceDetailsWidget::AckDayOrEnable() noexcept
 {
-  AirspaceDetailsWidget *widget =
-    new AirspaceDetailsWidget(airspace, warnings);
+  // Use specialized widget for NOTAMs
+  const bool is_notam = airspace->GetType() == AirspaceClass::NOTAM;
+  
+  AirspaceDetailsWidget *widget = is_notam
+    ? new NOTAMDetailsWidget(airspace, warnings)
+    : new AirspaceDetailsWidget(airspace, warnings);
+  
+  const char *title = is_notam ? _("NOTAM Details") : _("Airspace Details");
```
Imo this is fine now, the class/type thing was drift and should be gone now.

yorickreum commented on src/NOTAM/NOTAMGlue.cpp:
```diff
@@ -0,0 +1,1709 @@
+  return stats;
+}
+
+std::optional<struct NOTAM>
+NOTAMGlue::FindNOTAMByNumber(const std::string &number) const
+{
+  const std::lock_guard<Mutex> lock(mutex);
+  auto *impl = current_notams_impl.get();
+  
+  for (const auto &notam : impl->current_notams) {
+    if (notam.number == number) {
+      return notam; // Return a copy while holding the lock
```
Fixed?

yorickreum commented on NEWS.txt:
```diff
@@ -25,6 +25,11 @@ Version 7.45 - not yet released
+  - add support for downloading, caching and displaying NOTAM airspaces
+  - add manual refresh and auto-refresh with location/time based updates
+  - add NOTAM settings and filters (IFR, effective time, max radius, hidden Q-codes)
+  - add NOTAM details integration in airspace dialogs
```
There is no related issue currently afaik, only this PR, should we put it or is it okay? Or should we create an issue? @lordfolken

yorickreum commented on src/Dialogs/Settings/Panels/NOTAMConfigPanel.cpp:
```diff
@@ -0,0 +1,471 @@
+  char format_display[32], format_edit[32];
+  StringFormat(format_display, ARRAY_SIZE(format_display),
+               _("%%.0f %s"), unit_name);
+  std::strcpy(format_edit, "%.0f");
+  
+  AddFloat(_("Maximum NOTAM Radius"),
+           _("Filter out NOTAMs with radius larger than this. Set to 0 to disable."),
+           format_display, format_edit,
+           min_max_radius_user, max_max_radius_user, step_max_radius_user, 0,
+           max_radius_user);
+  StringFormat(buffer, ARRAY_SIZE(buffer),
+               _("%u filtered"), stats.filtered_by_radius);
+  AddReadOnly("", nullptr, buffer);
+
+  AddText(_("Hidden Q-Codes"),
+          _("Space-separated Q-code prefixes to hide (e.g., QA QK QN QOA QOL)."),
+          computer.notam.hidden_qcodes.c_str());
+  StringFormat(buffer, ARRAY_SIZE(buffer),
+               _("%u filtered"), stats.filtered_by_qcode);
+  AddReadOnly("", nullptr, buffer);
+
+  UpdateVisibility();
+#endif
+}
+
+void
+NOTAMConfigPanel::Show(const PixelRect &rc) noexcept
+{
+#ifdef HAVE_HTTP
+  // Register as listener for NOTAM updates
+  if (net_components && net_components->notam) {
+    net_components->notam->AddListener(*this);
+  }
+  
+  ConfigPanel::BorrowExtraButton(1, _("Refresh"), [this](){
+    OnUpdateButton();
+  });
+  
+  ConfigPanel::BorrowExtraButton(2, _("List"), [](){
+    ShowNOTAMListDialog(UIGlobals::GetMainWindow());
```
Yes we will do `make update-po` in a follow up PR. But this PR is complex enough, let's not add translations to it.

yorickreum commented:
@coderabbitai resume and review

yorickreum commented on src/NOTAM/NOTAMGlue.cpp:
```diff
@@ -0,0 +1,1738 @@
+      }
+      // NOTAMListener::OnNOTAMsLoadComplete() is noexcept by interface
+      for (auto *listener : listeners_copy)
+        listener->OnNOTAMsLoadComplete(false);
```
Valid concern, but still too broad for this initial feature PR. The current listener API is intentionally non-owning/raw-pointer based, so fixing this properly would require redesigning listener lifetime semantics rather than a localized NOTAM change.

yorickreum commented:
@corerabbitai full review

yorickreum commented on src/NOTAM/NOTAMGlue.cpp:
```diff
@@ -0,0 +1,1764 @@
+}
+
+void
+NOTAMGlue::SetSettings(const NOTAMSettings &_settings) noexcept
+{
+  const std::lock_guard<Mutex> lock(mutex);
+  settings = _settings;
+}
```
I think this is fine, edge case not worth adding complexity.

yorickreum commented on src/Renderer/AirspaceLabelRenderer.cpp:
```diff
@@ -83,9 +241,67 @@ AirspaceLabelRenderer::DrawInternal(Canvas &canvas,
+
+      const auto pos = projection.GeoToScreenIfVisible(airspace.GetCenter());
+      if (!pos)
+        continue;
+
+      const auto label = MakeNotamLabelText(airspace);
+
+      NotamLabelCluster *cluster = nullptr;
+      for (auto &candidate : clusters)
+        if (MatchesNotamCluster(candidate, *pos, cluster_distance)) {
+          cluster = &candidate;
+          break;
+        }
+
+      if (cluster == nullptr) {
+        if (clusters.full()) {
+#ifndef NDEBUG
+          LogFmt("AirspaceLabelRenderer: skipped NOTAM label at {},{} "
+                 "(cluster buffer full)",
+                 pos->x, pos->y);
+#endif
+          continue;
+        }
+
+        cluster = &clusters.append();
+        cluster->anchor = *pos;
+        cluster->count = 0;
+        cluster->labels.clear();
+      }
+
+      AddToNotamCluster(*cluster, label);
+    }
+
+    for (const auto &cluster : clusters)
+      DrawNotamCluster(canvas,
+                       AdjustNotamClusterAnchor(cluster.anchor,
+                                                projection.GetScreenRect()),
+                       cluster, mode,
+                       projection.GetScreenRect(), label_block);
+  }
```
Partly valid concern, but this would broaden the PR from NOTAM label handling into the global airspace altitude-label rendering path. Leaving it as-is only risks some visual overlap/inconsistency; it does not affect NOTAM correctness or airspace logic, so I’d keep that refactor out of this feature PR.

yorickreum commented on src/NOTAM/NOTAMGlue.hpp:
```diff
@@ -0,0 +1,289 @@
+  /** Mark the next successful load as user-initiated. */
+  void MarkManualRefreshRequested() noexcept;
+
+  void SetSettings(const NOTAMSettings &_settings) noexcept;
```
Not actionable here. These methods use XCSoar’s Mutex, and the codebase already treats `std::lock_guard<Mutex>` in `noexcept` methods as normal; I don’t want to special-case `SetSettings()` / `GetSettingsSnapshot()` based on a generic `std::mutex` concern.

yorickreum commented:
@coderabbitai resume and review

yorickreum commented on src/NOTAM/Client.cpp:
```diff
@@ -0,0 +1,680 @@
+        throw std::runtime_error("GeometryCollection has no usable geometry");
+      }
+
+      if (geom_type == "Point"sv) {
+        const auto coords_it = geometry.find("coordinates");
+        if (coords_it == geometry.end() || !coords_it->value().is_array())
+          throw std::runtime_error("Point missing coordinates");
+
+        const auto &coords = coords_it->value().as_array();
+        if (coords.empty())
+          throw std::runtime_error("Point has empty coordinates");
+
+        notam.geometry.type = NOTAM::NOTAMGeometry::Type::POINT;
+        notam.geometry.center =
+          boost::json::value_to<NOTAMPoint>(coords_it->value());
+        notam.geometry.radius_meters = 0.0;
+        notam.geometry.polygon_points.clear();
+        apply_radius_if_present();
+        return true;
```
Why would the rendered circle for a point NOTAM exceed the configured limit?

Commit:
test/notam: add TAP parser and filter coverage

yorickreum commented on test/src/TestNOTAM.cpp:
```diff
@@ -0,0 +1,156 @@
+int
+main()
+{
+  plan_tests(34);
```
`TestNOTAM` really has 34 ok1() assertions, hasn't it? So plan_tests(34) is already correct?

yorickreum commented:
Puuh, turns out, in 5.000 lines of code there can be a crazy amount of edge cases. Some of them actually might be relevant. I really hope iterating this with the rabbit makes it significantly more robust, it turned out to be quite some effort.

yorickreum commented:
@coderabbitai resume and review

yorickreum commented:
@coderabbitai full review

lordfolken commented:
## Airspace day acknowledgement (non-NOTAM)

During review of the `AirspaceWarningManager::AcknowledgeDay` changes for NOTAM persistence, there may be an interaction with **non-NOTAM** airspaces.

The new early return at the start of `AcknowledgeDay`:

```cpp
if (!set && key == nullptr && !was_member)
  return;
```

For any airspace that is **not** `AirspaceClass::NOTAM`, `NotamDayAckKey()` returns `nullptr`, so `was_member` is false. For `AcknowledgeDay(airspace, false)` (clear “acknowledged for today”), all three parts of the condition are true and the function **returns before** `warning->AcknowledgeDay(set)` runs.

On `master`, clearing always went through `GetWarning(...).AcknowledgeDay(set)`. One existing caller that passes `false` for a **generic** selected airspace is `AirspaceWarningListWidget::Enable()` in `dlgAirspaceWarnings.cpp` (`lease->AcknowledgeDay(airspace, false)` after re-enabling inside/warning ack).

**Suggestion:** Narrow or remove this early return so clearing day acknowledgement still runs for non-NOTAM airspaces (e.g. only apply the shortcut when NOTAM-specific logic applies, or gate it on `GetType() == AirspaceClass::NOTAM`). Adding a small test or manual check: acknowledge for day on a class C/D airspace, then use **Enable** — day ack should clear as before.

Happy to help double-check a patch if useful.

lordfolken commented:
## Off-by-one / bounds review (NOTAM PR)

Quick pass over NOTAM-related paths for indexing, token parsing, and buffer use.

### Looks consistent

- **`NOTAMList.cpp`**: `HEADER_COUNT` matches the four header rows; `is_header = (i < HEADER_COUNT)`; `SetLength(max(1, items.size()))` pairs with `OnPaintItem` handling `items.empty()` only for `i == 0`; `i >= items.size()` guard before `items[i]`.
- **`Filter.cpp` (`IsQCodeHidden`)**: Token lengths use half-open `[start, end)`; loop `i < prefix_len` with `hidden_list[start + i]` stays in range.
- **`Client.cpp` (polygon centroid)**: Duplicate closing vertex handled by decrementing `center_count` before the average loop.
- **`AirspaceLabelRenderer.cpp` (NOTAM clusters)**: When `cluster.count > 3`, `label_lines` is 2 and only `labels[0..1]` are read; overflow line uses `cluster.count > NOTAM_CLUSTER_VISIBLE_LINES` so `count - label_lines` is safe for the summary.
- **Cache file read**: `buffer[4096]` read loop appends `bytes_read` per iteration; no off-by-one NUL handling issue in the chunking itself.

### Minor footguns (no issue with current constants)

- **`NOTAMList.cpp` – `TruncateStringUTF8(text.c_str(), max_bytes - 3, max_bytes - 3)`**: The three-arg overload is `(max_chars, max_bytes)`; passing the same value for both is intentional for a rough “reserve 3 bytes for `...`” budget, but **if `max_bytes` were ever &lt; 3**, `max_bytes - 3` would **underflow** (`size_t`). With `constexpr max_bytes = 120` it is safe; consider `std::max(0u, max_bytes - 3)` or an assert if this is ever made configurable.

- **`DrawNotamCluster`**: `visible_lines` (for vertical centering) vs `label_lines` (what is drawn) can differ when `count > 3`—possible **layout** inconsistency, not an indexing overrun.

No further off-by-one issues stood out in these paths.

lordfolken commented:
## JSON load from cache file — possible reuse

`NOTAMGlue` reads cache files with a manual loop (`char buffer[4096]`, `FileReader::Read`, `std::string::append`) and then `boost::json::parse` on the full buffer (see e.g. `LoadCacheJsonValue` / `LoadNOTAMsFromFile`).

The tree already has **`Json::Parse(Reader &)`** in `src/json/Parse.cxx`: it streams from a `Reader` into `boost::json::stream_parser` (with the documented buffer-size workaround for boost/json).

**Suggestion (follow-up, not blocking):** consider wiring cache loads through `FileReader` + `Json::Parse`, or extracting a small **read-entire-file-to-string** helper if both paths must keep a single `std::string` for the same API. That would dedupe the chunked read logic and keep JSON parsing quirks in one place.

Happy to help sketch a minimal patch if useful.

lordfolken commented:
A couple of points:
 * my main condition is stability of existing functionality, so anything we touch that isn't strictly notam only has to be safe.
 * crashes should not happen
 * we should seprate any non-related changes. (i.e. ios build pipeline fixes)

lordfolken commented:
## Layering / `NOTAMGlue` split (follow-up)

`NOTAMGlue.cpp` currently combines several concerns that XCSoar’s layering guidelines usually keep apart:

- **Network + JSON/cache I/O** (curl, `NOTAMClient`, file cache, delta merge)
- **In-memory NOTAM state** (`NOTAMImpl`, known map, metadata)
- **Integration with the engine** (`UpdateAirspaces`, converter, terrain ground levels — callers already use `ScopeSuspendAllThreads` where needed)
- **UI-adjacent behaviour** (`Message::AddMessage`, `NOTAMListener` via `UI::Notify`)

That makes the class hard to review and easy to accidentally pull upper layers into fetch logic.

**Suggestion (post-merge or separate PR):** split along responsibility, e.g.

1. **Fetch/cache module** — load/save, HTTP coroutine, delta apply, `can_commit` / `mutation_generation`; **no** direct `Airspaces` / map UI includes.
2. **Thin airspace sync** — filter + `NOTAMConverter` + `Airspaces` update (or keep as clearly named free functions).
3. **Small `NOTAMGlue` facade** — owns `CurlGlobal`, timer/trigger entry points, wires fetch → sync → notify listeners / user messages.

Optional: isolate **cache file path + read/write** into a tiny `NOTAMCache` helper to dedupe with `Json::Parse` discussion elsewhere.

This is **not** a functional requirement for the feature itself; it’s **maintainability and alignment** with project layering. Happy to help sketch a minimal first extraction if useful.

yorickreum commented:
@coderabbitai full review

Commit:
test/airspace: add AckDay regression coverage

yorickreum commented:
@coderabbitai resume and full review

yorickreum commented:
@coderabbitai review

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@coderabbitai resume with full review

Commit:
Dialogs/MapItemListDialog: Harden Enable action handling

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@coderabbitai resume and review

yorickreum commented on src/NOTAM/Client.cpp:
```diff
@@ -0,0 +1,832 @@
+  }
+}
+
+static std::vector<NOTAM::NOTAMGeometry>
+ParseNOTAMGeometries(const NOTAM &base_notam,
+                     const boost::json::object &feature,
+                     const boost::json::object &notam_obj)
```
IMO that is okay for now. NOTAMs with multiple geometries are kinda rare. If they appear twice in the list, so be it.

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@coderabbitai full review

yorickreum commented on src/Dialogs/Settings/Panels/NOTAMConfigPanel.cpp:
```diff
@@ -0,0 +1,523 @@
+  RowFormWidget::Show(rc);
+}
+
+void
+NOTAMConfigPanel::Hide() noexcept
+{
+#ifdef HAVE_HTTP
+  // Unregister as listener
+  if (net_components && net_components->notam) {
+    net_components->notam->RemoveListener(*this);
+  }
+
+  notify.ClearNotification();
+  
+  ConfigPanel::ReturnExtraButton(1);
+  ConfigPanel::ReturnExtraButton(2);
+#endif
+
+  RowFormWidget::Hide();
+}
```
Already looks okay to me?

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@lordfolken I think this PR has undergone quite some hardening and fine tuning now and should be mostly safe to merge in

The only real concern for me is that it still by default uses my NOTAM proxy. What do we need to do to get this running on XCSoar infrastructure?

yorickreum commented:
@coderabbitai resume and review

yorickreum commented:
@coderabbitai explain my rate limits to me

yorickreum commented:
@coderabbitai reviews remaining?

yorickreum commented:
@coderabbitai reviews remaining?

yorickreum commented:
@coderabbitai reviews remaining?

yorickreum commented:
@coderabbitai reviews remaining?

lordfolken commented:
@yorickreum so i have bootstrapped the notam-proxy here: https://notam.xcsoar.org/ I'm waiting for the FFA credentials.

yorickreum commented:
@coderabbitai full review

yorickreum commented on test/src/TestAirspaceWarningManager.cpp:
```diff
@@ -0,0 +1,141 @@
+int
+main()
+{
+  plan_tests(31);
```
Bro, you keep counting wrong.

yorickreum commented:
@coderabbitai resume & review

yorickreum commented:
@coderabbitai full review

lordfolken commented:
@yorickreum https://notam.xcsoar.org/ should be working now.

yorickreum commented:
@coderabbitai any final remarks? ready for human review and merge?

yorickreum commented:
@coderabbitai rate limit?
Can I use my personal CodeRabbit subscription here?

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@coderabbitai full review

yorickreum commented:
@coderabbitai resume and review

yorickreum commented:
> 
> 
> <details>
> <summary>♻️ Duplicate comments (1)</summary><blockquote>
> 
> <details>
> <summary>test/src/TestNOTAM.cpp (1)</summary><blockquote>
> 
> `70-70`: _⚠️ Potential issue_ | _🟡 Minor_ | _⚡ Quick win_
> 
> **`plan_tests()` count is higher than executed assertions.**
> 
> `plan_tests(54)` does not match the current number of `ok1()` calls in this file (50), so the test will fail on TAP plan mismatch even when logic is correct.
> 
> <details>
> <summary>✅ Minimal fix</summary>
> 
> ```diff
> -  plan_tests(54);
> +  plan_tests(50);
> ```
> </details>
> 
>   
> As per coding guidelines, "Use plan_tests(N) to declare the total assertion count upfront, and ensure it exactly matches the total number of ok1() calls in the test file."
> 
> <details>
> <summary>🤖 Prompt for AI Agents</summary>
> 
> ```
> Verify each finding against current code. Fix only still-valid issues, skip the
> rest with a brief reason, keep changes minimal, and validate.
> 
> In `@test/src/TestNOTAM.cpp` at line 70, The TAP plan count in plan_tests(54) does
> not match the actual number of ok1() assertions (50); update the plan to match
> by changing the call to plan_tests(50) or alternatively add the four missing
> ok1() assertions so the total ok1() calls equals the plan; locate the
> plan_tests(...) invocation and the ok1(...) assertions in TestNOTAM.cpp to make
> the counts consistent.
> ```
> 
> </details>
> 
> </blockquote></details>
> 
> </blockquote></details>
> 
> <details>
> <summary>🤖 Prompt for all review comments with AI agents</summary>
> 
> ```
> Verify each finding against current code. Fix only still-valid issues, skip the
> rest with a brief reason, keep changes minimal, and validate.
> 
> Duplicate comments:
> In `@test/src/TestNOTAM.cpp`:
> - Line 70: The TAP plan count in plan_tests(54) does not match the actual number
> of ok1() assertions (50); update the plan to match by changing the call to
> plan_tests(50) or alternatively add the four missing ok1() assertions so the
> total ok1() calls equals the plan; locate the plan_tests(...) invocation and the
> ok1(...) assertions in TestNOTAM.cpp to make the counts consistent.
> ```
> 
> </details>
> 
> ---

Sure, and how many e are there in Meerrettich, @coderabbitai?

yorickreum commented on src/NOTAM/Converter.cpp:
```diff
@@ -0,0 +1,229 @@
+      top.reference = AltitudeReference::MSL;
+    }
+    
+    // Keep human-readable NOTAM content in name and stable identifier in
+    // station_name for lookups/dialog details.
+    std::string name = !notam.text.empty()
+      ? notam.text
+      : (!notam.number.empty() ? notam.number : ("NOTAM " + notam.id));
+    std::string station_name = !notam.number.empty()
+      ? notam.number
+      : notam.id;
```
Current `FindNOTAMByNumber()` already searches both `notam.number` and `notam.id`, so the current `station_name = number-or-id` does not break `dlgAirspaceDetails()` lookup. Removing the `id` fallback would require a broader representation change for numberless NOTAMs and is not needed to fix the cited failure?

yorickreum commented on src/Dialogs/Airspace/NOTAMList.cpp:
```diff
@@ -0,0 +1,368 @@
+    NOTAMStruct header1, header2, header3, header4;
+    
+    // Last Update time
+    std::time_t last_update = snapshot.last_update_time;
+    header1.number = _("Last Update");
+    if (last_update > 0) {
+      struct tm tm_buf;
+#ifdef _WIN32
+      const auto *tm =
+        (localtime_s(&tm_buf, &last_update) == 0) ? &tm_buf : nullptr;
+#else
+      const auto *tm = localtime_r(&last_update, &tm_buf);
+#endif
+      if (tm != nullptr) {
+        char time_buffer[64];
+        StringFormat(time_buffer, sizeof(time_buffer),
+                     "%04d-%02d-%02d %02d:%02d",
+                     tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+                     tm->tm_hour, tm->tm_min);
+        header1.text = time_buffer;
```
Fixed?

yorickreum replied:
There is an explicit "Last Update (local)" label right above?

lordfolken committed:
Engine/Airspace: Share OpenAir altitude limit parsing

Move altitude token parsing from AirspaceParser into AirspaceAltitude so
OpenAir and upcoming NOTAM callers use one implementation. OpenAir keeps
lenient unknown-token handling and the historical 50000 m UNL ceiling.

lordfolken committed:
NOTAM/Client: Use shared ParseAirspaceAltitude

Wire NOTAM JSON altitude strings through Engine/Airspace parsing with
strict tokens, AMSL, and the NOTAM unlimited ceiling in feet. Extend
TestNOTAM to link libairspace.

lordfolken commented:
Closing in favor of a clean replacement PR after rebasing and squashing follow-up fixes (NOTAM module refactor, load deadlock, TIM/GeoJSON `GeoPoint` parser conflict, and extended tests).

Please use **#2540** (`lordfolken:notam-integration`) for review going forward.
